plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryQueryPageSource.java

```diff
@@ -0,0 +1,248 @@
+import io.trino.spi.type.VarbinaryType;
+import io.trino.spi.type.VarcharType;
+
+import java.math.BigDecimal;
+import java.time.LocalDate;
+import java.time.LocalTime;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeFormatterBuilder;
+import java.util.List;
+import java.util.Optional;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Verify.verify;
+import static io.airlift.slice.Slices.utf8Slice;
+import static io.trino.plugin.bigquery.BigQueryType.toTrinoTimestamp;
+import static io.trino.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
+import static io.trino.spi.type.BigintType.BIGINT;
+import static io.trino.spi.type.DateType.DATE;
+import static io.trino.spi.type.Decimals.isLongDecimal;
+import static io.trino.spi.type.IntegerType.INTEGER;
+import static io.trino.spi.type.TimeType.TIME_MICROS;
+import static io.trino.spi.type.TimestampType.TIMESTAMP_MICROS;
+import static io.trino.spi.type.Timestamps.NANOSECONDS_PER_DAY;
+import static io.trino.spi.type.Timestamps.PICOSECONDS_PER_DAY;
+import static io.trino.spi.type.Timestamps.PICOSECONDS_PER_NANOSECOND;
+import static io.trino.spi.type.Timestamps.round;
+import static java.lang.String.format;
+import static java.time.temporal.ChronoField.NANO_OF_SECOND;
+import static java.util.Objects.requireNonNull;
+
+public class BigQueryQueryPageSource
+        implements ConnectorPageSource
+{
+    private static final DateTimeFormatter TIME_FORMATTER = new DateTimeFormatterBuilder()
+            .appendPattern("HH:mm:ss")
+            .optionalStart()
+            .appendFraction(NANO_OF_SECOND, 0, 6, true)
+            .optionalEnd()
+            .toFormatter();
+
+    private final List<Type> columnTypes;
+    private final PageBuilder pageBuilder;
+    private final TableResult tableResult;
+
+    private long count;
+    private boolean finished;
+
+    public BigQueryQueryPageSource(
+            BigQueryClient client,
+            BigQueryTableHandle table,
+            List<String> columnNames,
+            List<Type> columnTypes,
+            Optional<String> filter)
+    {
+        requireNonNull(client, "client is null");
+        requireNonNull(table, "table is null");
+        requireNonNull(columnNames, "columnNames is null");
+        requireNonNull(columnTypes, "columnTypes is null");
+        requireNonNull(filter, "filter is null");
+        checkArgument(columnNames.size() == columnTypes.size(), "columnNames and columnTypes sizes don't match");
+        this.columnTypes = ImmutableList.copyOf(columnTypes);
+        this.pageBuilder = new PageBuilder(columnTypes);
+        TableId tableId = TableId.of(client.getProjectId(), table.getRemoteTableName().getDatasetName(), table.getRemoteTableName().getTableName());
+        this.tableResult = client.query(tableId, ImmutableList.copyOf(columnNames), filter);
+    }
+
+    @Override
+    public long getCompletedBytes()
+    {
+        return count;
```

Comment:
this is returning number of rows read instead of bytes. Is this expected?

Reply:
My bad. Changed to 0.